### PR TITLE
upgrade: restrict cancelation of the upgrade

### PR DESCRIPTION
**Pull request by MaximilianMeister**
after certain steps it shouldn't be allowed to cancel the upgrade
anymore

### CODE REVIEW

File: crowbar_framework/lib/crowbar/error/upgrade_cancel.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+module Crowbar
+  module Error
+    class UpgradeCancelError < StandardError
+      def initialize(step_name = "")
+        super("Not able to cancel the upgrade at the step #{step_name}")
```
Comment (jsuchome) [inline crowbar_framework/lib/crowbar/error/upgrade_cancel.rb]: I'd say "Not possible to cancel..."

**Comment by jsuchome** [comment]
There's still something wrong here...

```
crowbarctl upgrade cancel
```
leads to 
```
E, [2016-12-09T17:01:43.022551 #6734:0x00000009134fd0] ERROR -- Not able to cancel the upgrade at the step nodes_repo_checks
I, [2016-12-09T17:01:43.024193 #6734:0x00000009134fd0]  INFO -- Completed 500 Internal Server Error in 4ms (ActiveRecord: 0.0ms)
F, [2016-12-09T17:01:43.025235 #6734:0x00000009134fd0] FATAL -- 
NameError (uninitialized constant Crowbar::Error::UpgradeCancelError):
  app/controllers/api/upgrade_controller.rb:233:in `rescue in cancel'
  app/controllers/api/upgrade_controller.rb:218:in `cancel'
```

**Comment by MaximilianMeister** [comment]
> There's still something wrong here...

should be fixed now

File: crowbar_framework/app/controllers/api/upgrade_controller.rb [hunk 2]
```diff
@@ -228,6 +229,24 @@ def cancel
         }
       }, status: cancel_upgrade[:status]
     end
+  rescue Crowbar::Error::UpgradeCancelError => e
```
Comment (skazi0) [inline crowbar_framework/app/controllers/api/upgrade_controller.rb]: Why did you move this from model to controller? There are still some errors returned via `cancel_upgrade` and some are handled as exceptions. Why not keep it consistent?

Comment (MaximilianMeister) [inline crowbar_framework/app/controllers/api/upgrade_controller.rb]: the exceptions from `revert_nodes_from_crowbar_upgrade` will be caught in the `rescue StandardError` block. and what exactly do you think is inconsistent here?

Comment (MaximilianMeister) [inline crowbar_framework/app/controllers/api/upgrade_controller.rb]: i see, i'll clean that up a bit. thanks for the hint

File: crowbar_framework/lib/crowbar/upgrade_status.rb [hunk 3]
```diff
@@ -129,6 +129,16 @@ def finished?
       current_step == upgrade_steps_6_7.last
     end
 
+    def cancel_allowed?
+      [
+        :upgrade_prechecks,
+        :upgrade_prepare,
+        :admin_backup,
+        :admin_repo_checks,
+        :admin_upgrade
```
Comment (MaximilianMeister) [inline crowbar_framework/lib/crowbar/upgrade_status.rb]: i have solved it now by adding an additional check if the `admin_upgrade` is running

File: crowbar_framework/app/controllers/api/upgrade_controller.rb [hunk 4]
```diff
@@ -213,21 +213,38 @@ def prechecks
...
     else
       render json: {
         errors: {
           cancel: {
-            data: cancel_upgrade[:message],
+            data: I18n.t("api.upgrade.cancel.failed"),
             help: I18n.t("api.upgrade.cancel.help.default")
           }
         }
       }, status: cancel_upgrade[:status]
```
Comment (skazi0) [inline crowbar_framework/app/controllers/api/upgrade_controller.rb]: Where is this `cancel_upgrade` variable initialized now?

**Comment by MaximilianMeister** [comment]
rebased
